oak_restricted_kernel/src/payload.rs

```diff
@@ -128,33 +133,112 @@ impl Application {
 
-        log::info!("Running application");
+        self.entry()
+    }
+}
+
+pub struct Process {
+    pml4: x86_64::structures::paging::PageTable,
+    entry: VirtAddr,
+}
+
+impl Process {
+    /// Creates a process from the application, without executing it.
+    ///
+    /// # Safety
+    ///
+    /// The application must be built from a valid ELF file representing an Oak Restricted
+    /// Application.
+    pub unsafe fn from_application(application: &Application) -> Result<Self, anyhow::Error> {
+        let pml4 = crate::BASE_L4_PAGE_TABLE
+            .get()
+            .context("base l4 table should be set")?
+            .clone();
+        // Load the process's page table, so the application can be loaded into its memory. Hold
+        // onto the previous PT, so we can revert to it once the application has been mapped
+        // into the process pt.
+        let mut outer_prev_page_table = {
+            let pml4_frame: x86_64::structures::paging::PhysFrame = {
+                let phys_addr = {
+                    let addr = core::ptr::addr_of!(pml4);
+                    let pt_guard = crate::PAGE_TABLES.lock();
+                    let pt = pt_guard.get().expect("failed to get page tables");
+                    crate::mm::Translator::translate_virtual(pt, VirtAddr::from_ptr(addr))
+                        .expect("failed to translate virtual page table address")
+                };
+                x86_64::structures::paging::PhysFrame::from_start_address(phys_addr)
+                    .expect("couldn't get the physical frame for page table address")
+            };
+            // Safety: the new page table maintains the same mappings for kernel space.
+            unsafe { crate::PAGE_TABLES.lock().replace(pml4_frame) }
+                .expect("at this point there should be a previous root pt")
+                .into_inner()
+        };
+
+        // Safety: the application must be valid.
+        let entry = unsafe { application.map_into_memory() };
+
+        // We've mapped the memory into the process page tables. Let's revert to the previous page
+        // table.
+        {
+            let mut mapped_prev_pt = outer_prev_page_table.inner().lock();
+            let prev_page_table = mapped_prev_pt.level_4_table();
+            let pml4_frame: x86_64::structures::paging::PhysFrame = {
+                let phys_addr = {
+                    let addr = core::ptr::addr_of!(*prev_page_table);
+                    let pt_guard = crate::PAGE_TABLES.lock();
+                    let pt = pt_guard.get().expect("failed to get page tables");
+                    crate::mm::Translator::translate_virtual(pt, VirtAddr::from_ptr(addr))
+                        .expect("failed to translate virtual page table address")
+                };
+                x86_64::structures::paging::PhysFrame::from_start_address(phys_addr)
+                    .expect("couldn't get the physical frame for page table address")
+            };
```

Comment:
Okay so this is a tricky one. The logic to identify the frame holding pml4 for repeated a few times throughout this PR. — I tried extracting it into a function that accepts `&PagetTable`, but it a false addr being identified.

```

pub fn identify_pml4_frame(pml4: &PageTable) -> PhysFrame {
    let phys_addr = {
        let addr = core::ptr::addr_of!(*pml4);
        let pt_guard = crate::PAGE_TABLES.lock();
        let pt = pt_guard.get().expect("failed to get page tables");
        pt.translate_virtual(VirtAddr::from_ptr(addr))
            .expect("failed to translate virtual page table address")
    };
    PhysFrame::from_start_address(phys_addr)
        .expect("couldn't get the physical frame for page table address")
}
```

Maybe it was just late on Friday, but for now I decided to just duplicate this code around as is.

Reply:
I suspect the issue is with addr_of, perhaps it is trying to get the address of the reference passed to this function which is on the stack? I always forget the correct syntax for addr_of, perhaps try adding or removing more indirection (`*`) until it works :)

Reply:
> I don't think you even need `addr_of` if you have a reference. You should be able to just directly coerce the reference to a pointer (something like `let addr = pml4 as *const PageTable;` or `let addr = &*pml4 as *const PageTable;`).

this worked fine, thanks!